### PR TITLE
Add presigned url feature to the minio client

### PR DESCRIPTION
**Pull request by enescakir**
### Refactor Minio::Client to enhance compatibility with Python version
Furkan ported the Ruby version of "Minio::Client" from the Python version https://github.com/minio/minio-py.

I will migrate more endpoints from it. Minor refactoring was done to make the method interfaces more similar.

### Add presigned url feature to the minio client

We will download large boot images from our MinIO cluster, replacing Azure Blob Storage. Currently, we generate download URLs using SAS tokens from Azure.

MinIO offers similar functionality through presigned URLs. These URLs eliminate the need to send MinIO credentials to VM hosts. They expire after a specified period, making them safe for dataplane transmission. We can also download the presigned URL using any client, including "curl"

I ported it from Python method.

https://github.com/minio/minio-py/blob/fbcfca1b96c8919a496f8420dba0be139091364b/minio/api.py#L2216

### CODE REVIEW

File: lib/minio/header_signer.rb [hunk 1]
```diff
@@ -16,22 +16,22 @@
 class Minio::HeaderSigner
   SERVICE_NAME = "s3"
   def build_headers(method, uri, body, creds, region)
+    date = Time.now.utc
```
Comment (furkansahin) [inline lib/minio/header_signer.rb:19]: since we're refactoring, let's name this `timestamp`, it makes more sense.

File: lib/minio/client.rb [hunk 2]
```diff
@@ -77,6 +77,10 @@ def admin_policy_remove(policy_name)
     response.status
   end
 
+  def get_presigned_url(method, bucket_name, object_name, expires: 3600, request_date: Time.now.utc)
```
Comment (furkansahin) [inline lib/minio/client.rb:80]: does this `request_date` allow creating a URL from an earlier time or a URL that will be valid in future? I guess the latter is OK but the first one, not so much. Also, that functionality doesn't exist in UI. Should we simply move `request_date` into the function and not allow in the API?

**Review by furkansahin — APPROVED** [review]
Hope there were enough examples in the code base already. I know for a fact that this client translation is painful. Thanks a lot!

File: lib/minio/header_signer.rb (hunk 1 above)
Comment (enescakir) [reply lib/minio/header_signer.rb:19]: The Python version used 'date'. If you're not that strong, I can keep them the same. This simplifies porting other functions.

https://github.com/minio/minio-py/blob/fbcfca1b96c8919a496f8420dba0be139091364b/minio/api.py#L241

File: lib/minio/client.rb (hunk 2 above)
Comment (enescakir) [reply lib/minio/client.rb:80]: I used the same parameters with the Python library. If you believe it's unnecessary, I can remove it.

Comment (furkansahin) [reply lib/minio/client.rb:80]: Let's move it into thw function for now. If we need a url to be valid in feature, we can always move it back. I don't see a reason to introduce it into the API, for now.

Comment (enescakir) [reply lib/minio/client.rb]: I moved it

**Review by fdr — COMMENTED** [review]
You use an hour expiration, they use seven days, should expiry not be defaulted perhaps?

**Comment by enescakir** [comment]
> You use an hour expiration, they use seven days, should expiry not be defaulted perhaps?

Makes sense. To prevent confusion, I've removed the default value.

**Review by fdr — APPROVED** [review]

**Pull request merged**